templates/install_puppet.ps1.erb

```diff
@@ -0,0 +1,185 @@
+    }
+  }
+}
+
+# ************** Execution start **************
+$ErrorActionPreference = "Stop"
+$service_names=@(
+  "puppet",
+  "pxp-agent",
+  "mcollective"
+)
+$msi_arguments = "/qn /norestart /i `"$Source`" /l*vx+ `"$Logfile`" $InstallArgs"
+
+if (Test-Path "HKLM:\SOFTWARE\Puppet Labs") {
+  $msi_arguments += " ENVIRONMENT=`"$(puppet.bat config print --section agent environment)`""
+}
+if ($InstallDir) {
+  $msi_arguments += " INSTALLDIR=`"$InstallDir`""
+}
+if ($PuppetMaster) {
+  $msi_arguments += " PUPPET_MASTER_SERVER=`"$PuppetMaster`""
+}
+if ($PuppetStartType) {
+  $msi_arguments += " PUPPET_AGENT_STARTUP_MODE=`"$PuppetStartType`""
+}
+Out-File -InputObject $PID -FilePath $InstallScriptPIDFile
+try {
+  # Wait for the puppet run to finish
+  Out-File -InputObject "Waiting for puppet to stop, PID:$PuppetPID" -FilePath $Logfile -Append
+  try {
+    Wait-Process -Id $PuppetPID -Timeout 120
+  # We have to catch all exceptions here since the Error action preference of 'stop' will cause wait-process
+  # to actually throw [System.Management.Automation.ActionPreferenceStopException] rather than the actual
+  # exceptions. However, once inside the catch block the value of $_ is correctly set as the actual exception
+  } catch {
+    if ($_.Exception.GetType() -eq [Microsoft.PowerShell.Commands.ProcessCommandException]) {
+      Out-File -InputObject "Puppet was already finished, continuing..." -FilePath $Logfile -Append
+    } elseif ($_.Exception.GetType() -eq [System.TimeoutException]) {
+      Out-File -InputObject "ERROR: timed out waiting for puppet run to finish!" -FilePath $Logfile -Append
+      throw $_
+    } else {
+      Out-File -InputObject "ERROR: failed while trying to wait for puppet process with $($_.Exception.GetType())" -FilePath $Logfile -Append
+      throw $_
+    }
+  }
+  foreach($service in $service_names) {
+    $serv_exists = Get-Service -Name $service -ErrorAction SilentlyContinue
+    if ($serv_exists) {
+      Out-File -InputObject "Stopping $($service) before upgrade" -FilePath $Logfile -Append
+      Stop-Service $service
+    }
+  }
+  if ($UseLockedFilesWorkaround) {
+    $temp_puppetres = move_puppetres_dll
+  }
+  Out-File -InputObject "Beginning MSI installation with Arguments: $msi_arguments" -FilePath $Logfile -Append
+  Out-File -InputObject "****************************** Begin msiexec.exe output ******************************" -FilePath $Logfile -Append
+  $msiexec_proc = Start-Process -FilePath msiexec.exe -PassThru -ArgumentList $msi_arguments
+  # Using the -Wait option in Start-Process will wait on the process group instead of just the one msiexec process in older versions of windows.
+  # but using WaitForExit works fine.
+  $msiexec_proc.WaitForExit()
+  Out-File -InputObject "****************************** End msiexec.exe output ******************************" -FilePath $Logfile -Append
+  if ($msiexec_proc.ExitCode -ne 0){

```

Comment:
Not quite true.

Exit code could be 0, 3010 or 1604 (? I think it's 1604 - Soft reboot required)

Reply:
Since we force /norestart in the MSI wouldn't the restart exit codes technically be failures as far as we're concerned?

It's my understanding that the agent should absolutely never cause a restart on upgrade.

Reply:
> Those are both successful installation methods. 1641 simply means "The requested operation completed successfully. The system will be restarted so the changes can take effect" and 3010 means "The requested operation is successful. Changes will not be effective until the system is rebooted."
>
> Thus one is initiating the reboot automatically, and the other is waiting for the user to reboot. The difference is in the switches or properties you've set when calling the MSI.

3010 is a succesfull install but it's pending a reboot.  i.e. the agent package WANTED to reboot but it couldn't due to /norestart